keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,472 @@
+With `node-affinity-eviction`, the only thing I need to do is to relabel the node, and `node-affinity-eviction` controller will do the rest.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern for this controller may be the overhead of the nodes with a large number of pods. 
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+However, the impact won't be significant due to the limitation on the number of pods per node.
+The evictions caused by this may be expensive depending on the number of pods and the requests conflicts and blocking PDBs.
+
+During the eviction, some pods may fail to be evicted. Users can use `pod_failed_eviction_node_affinity_total` metric to find them. 
+Typically, the reason for a failed eviction is due to a violation of the PDB. However, it is also possible that there could be issues with Kubernetes components. 
+In the case of the former, users need to manually check the relevant settings to ensure they are in compliance with the PDB. 
+If it is the latter, users should investigate the status of various Kubernetes components.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+	// If the affinity requirements specified by this field are not met at
+	// scheduling time, the pod will not be scheduled onto the node.
+	// If the affinity requirements specified by this field cease to be met
+	// at some point during pod execution (e.g. due to an update), the system
+	// will try to eventually evict the pod from its node.
+  // For v1alpha1, evictions will be handled by the Eviction API, 
+  // and will take the PDBs into consideration.
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node(excluding mirror pods), checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
+
+- For alpha stage, we'll use a simple eviction API to evict Pods.
+  - It will use Eviction API to evict the pod(The API takes the PDBs into consideration, so we do not have to use the PDBs explicitly here). 
+  - If eviction fails, we will take different actions based on the failure reasons:
+    - If eviction fails due to a violation of the PDB, we will keep the target Pod unchanged. 
+    - If eviction fails for any other reason, we will attempt a certain number of retries. If the eviction continues to fail after multiple retries, we will keep the target Pod unchanged. 
```

Comment:
same here? What does this mean? How and why do we stop retrying?

Reply:
> keep the target Pod unchanged.  

This means that we will not evict this pod in the current eviction round.

> How and why do we stop retrying?

If eviction api returns a non-200/429 status multiple times, it is highly likely that there is a configuration error or an error in a specific component, and we should not continue with futile attempts?

Reply:
- it would be good to distinguish what is an eviction round and how often/when that happens
- I do not think it is reasonable to do retries in one eviction round

Reply:
> I do not think it is reasonable to do retries in one eviction round

OK, so we'll skip the pod as long as eviction api returns a non-200 stauts?

Reply:
> it would be good to distinguish what is an eviction round and how often/when that happens

Added.